File: 0953-verifying-an-alien-dictionary/0953-verifying-an-alien-dictionary.cpp
class Solution {
private:
    bool checkCharRank(const string& a, const string& b, unordered_map<char, int>& charRank){
        int minLen = min(a.length(), b.length());
        for(int i = 0; i < minLen; i++){
            if(a[i] != b[i]){
                return charRank[a[i]] < charRank[b[i]];
            }
        }
        return a.length() <= b.length();
    }
public:
    bool isAlienSorted(vector<string>& words, string order) {
        //what to return if vector is empty or has 1 string
        if(words.size()<=1) return true;
        
        unordered_map<char, int> charRank;
        for(int i = 0; i < order.length(); i++){
            charRank[order[i]] = i;
        }
        
        for(int i = 1; i < words.size(); i++){
            if(!checkCharRank(words[i-1], words[i], charRank)){
                return false;
            }
        }
        return true;
    }
};

File: 0953-verifying-an-alien-dictionary/0953-verifying-an-alien-dictionary.py
def compareWord(word1, word2, alphabet):
        len1 = len(word1)
        len2 = len(word2)
        for i in range(min(len1, len2)):
            if alphabet[word1[i]] > alphabet[word2[i]]:
                return False
            if alphabet[word1[i]] < alphabet[word2[i]]:
                return True
        return len1 <= len2 #if same return shorter word example 'apple' vs 'app'
        
class Solution(object):
    def isAlienSorted(self, words, order):
        """
        :type words: List[str]
        :type order: str
        :rtype: bool
        """
        """
        requirements:
        blank charachters are less than anything character 
        shorter words with same letters go in front of larger words
        
        will I have to deal with speciial characters?
        the order will always be 26 letters?
        
        get a ranking for each letter in the order
        compare if the next word has a ranking less than or equal
        check the lenghth as well
        
        hashmap: {char : number}
        
        """
        if(len(words) <= 1):
            return True
        
        alphabet = {char : i for i, char in enumerate(order)}

        
        for i in range(len(words)-1):
            if not compareWord(words[i], words[i+1], alphabet):
                return False
            
        return True

        